cc/minigui_player.h

```diff
@@ -0,0 +1,152 @@
+    // main line. Children at index 1 and later are variations from the main
+    // line.
+    std::vector<AuxInfo*> children;
+
+    // Any SGF comments associated with this position.
+    std::string comment;
+  };
+
+  // If waiting for the opponent to play, consider thinking for a bit.
+  // Returns true if we pondered.
+  void Ponder() override;
+
+  Response HandleCmd(const std::string& line) override;
+  Response HandleClearBoard(CmdArgs args) override;
+  Response HandleGenmove(CmdArgs args) override;
+  Response HandleLoadsgf(CmdArgs args) override;
+  Response HandlePlay(CmdArgs args) override;
+
+  Response HandleEcho(CmdArgs args);
+  Response HandleInfo(CmdArgs args);
+  Response HandlePlaysgf(CmdArgs args);
+  Response HandlePruneNodes(CmdArgs args);
+  Response HandleReportSearchInterval(CmdArgs args);
+  Response HandleSelectPosition(CmdArgs args);
+  Response HandleVariation(CmdArgs args);
+  Response HandleWinrateEvals(CmdArgs args);
+
+  // Shared implementation used by HandleLoadsgf and HandlePlaysgf.
+  Response ProcessSgf(const std::vector<std::unique_ptr<sgf::Node>>& trees);
+
+  // Writes the search data for the tree search being performed at the given
+  // root to stderr. If leaf is non-null, the search path from root to leaf
+  // is also written.
+  void ReportSearchStatus(MctsNode* root, MctsNode* leaf);
+
+  // Writes the position data for the node to stderr as a JSON object.
+  void ReportPosition(MctsNode* node);
+
+  // Registers the given node as having been played during the game,
+  // assigning the node a unique ID and constructing AuxInfo for it.
+  AuxInfo* RegisterNode(MctsNode* node);
+
+  // Gets the AuxInfo for the given node.
+  // CHECK fails if there isn't any AuxInfo, which means that RegisterNode
+  // hasn't previously been called: this node doesn't correspond to a move
+  // played during the game or a variation (it's a node from tree search).
+  AuxInfo* GetAuxInfo(MctsNode* node) const;
+
+  // Clears the to_eval_ win rate evaluation queue and repopulates it.
+  void RefreshPendingWinRateEvals();
+
+  // Controls which variation is reported during tree search.
+  // If child_variation_ == Coord::kInvalid, the principle variation from the
+  // root is reported. Otherwise, the principle variation of the
+  // corresponding child of the root is reported.
+  Coord child_variation_ = Coord::kInvalid;
+
+  // Map from MctsNode to auxiliary info about that node used by the GtpPlayer.
+  absl::flat_hash_map<MctsNode*, std::unique_ptr<AuxInfo>> node_to_info_;
+
+  // Map from unique ID associated with every position played in a game or
+  // variation to the position's auxiliary info and MctsNode.
+  absl::flat_hash_map<std::string, AuxInfo*> id_to_info_;
+
+  // Queue of positions that require their win rate to be evaluated.
+  std::deque<AuxInfo*> to_eval_;
+
+  // Number of times to perform tree search for each position when evaluating
+  // its win rate.
+  int num_eval_reads_ = 6;
```

Comment:
6?

Reply:
With v14, the predicted win rate was still changing noticeably after 4 reads per move, so I increased it a bit. 8 reads per position took a little too long on my GPU, so I split the difference and went with 6.